File: 프로그래머스/2/42578. 의상/의상.py
'''
1.아이디어
    - clothes[i][1]을 key로 딕셔너리를 생성한다
    - key값 별로 나누어서 조합의 개수 추가한다
        - 조합의 종류
        1) for k in key 여기서 1개씩, 2개씩, ... , len(key)개씩 고르는 조합
'''
from collections import defaultdict

def solution(clothes):
    
    d = defaultdict(list) # list를 value값으로 갖는 빈 딕셔너리를 만듦

    for i in clothes:
        d[i[1]].append(i[0]) # 초기화 없이 바로 값을 추가할 수 있음
        
    result = 1
    for i in d.values(): #key의 개수만큼 서로다른 조합을 가질 수 있음 -> len(keys())를 모두 곱함
        result *= (len(i)+1) #각각 하나씩 모두 입는 경우 + 안입는 경우
        
    return result -1

File: 프로그래머스/2/42578. 의상/의상.java
/*
[아이디어]
1. 의상 종류를 key로, 의상의 개수를 value로 저장한다 - O(n)
2. 경우의 수 구하기: 의상은 같은 종류를 중복선택 할 수 없으므로 독립적임 (각각의 개수 곱해서 -1) - O(n)
*/
import java.util.*;

class Solution {
    public int solution(String[][] clothes) {
        // 1. HashMap 초기화
        Map<String, Integer> map = new HashMap<>();
        for(String[] cloth: clothes){
            String type = cloth[1];
            map.put(type, map.getOrDefault(type,0)+1);            
        }
        
        // 2. 경우의 수 계산하기 (안 입는 경우 +1 포함)
        int answer = 1;
        for (int count : map.values()) {
            answer *= (count + 1); // (해당 종류의 개수 + 안 입는 경우)
        }
        
        // 3. 아무것도 안 입는 경우 1가지를 빼고 반환
        return answer - 1;
        
    }
}